python/pyarrow/io.pxi

```diff
@@ -1424,6 +1430,18 @@ cdef class Buffer(_Weakrefable):
         are_equal : bool
             True if buffer contents and size are equal
         """
+        if self.device != other.device:
+            raise ValueError(
+                "Device on which the data resides differs between buffers: "
+                f"{self.device.type_name} and {other.device.type_name}."
+            )
+        if not self.is_cpu and not other.is_cpu:
```

Comment:
```suggestion
        if not self.is_cpu:
```

(because you now already checked the devices are equal, just checking for `self.is_cpu` should be sufficient)

Reply:
I rearranged a bit locally - think it makes more sense:

```python
if not self.is_cpu or not other.is_cpu:
  if self.device != other.device or self.address != other.address:
      raise NotImplementedError(
          "Implemented only for data on CPU device or data with equal "
          "addresses"
      )
```

can push if not too late.

Reply:
I think the current error message could actually be clearer? i.e. if you are comparing a buffer on two different devices it will explicitly say that that is not supported because they are on a different device

Reply:
Ok, great! 👍